terraform/eval_diff_test.go

```diff
@@ -0,0 +1,78 @@
+package terraform
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestEvalFilterDiff(t *testing.T) {
+	ctx := new(MockEvalContext)
+
+	cases := []struct {
+		Node   *EvalFilterDiff
+		Input  *InstanceDiff
+		Output *InstanceDiff
+	}{
+		// With no settings, it returns an empty diff
+		{
+			&EvalFilterDiff{},
+			&InstanceDiff{Destroy: true},
+			&InstanceDiff{},
+		},
+
+		// Destroy
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{Destroy: false},
+			&InstanceDiff{Destroy: false},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{Destroy: true},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Destroy: true,
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{},
+				},
+			},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{
+						RequiresNew: true,
+					},
+				},
+			},
+			&InstanceDiff{Destroy: true},
+		},
+		{
+			&EvalFilterDiff{Destroy: true},
+			&InstanceDiff{
+				Attributes: map[string]*ResourceAttrDiff{
+					"foo": &ResourceAttrDiff{},
+				},
+			},
+			&InstanceDiff{Destroy: false},
+		},
+	}
+
+	for i, tc := range cases {
+		var output *InstanceDiff
+		tc.Node.Diff = &tc.Input
+		tc.Node.Output = &output
+		if _, err := tc.Node.Eval(ctx); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+
+		if !reflect.DeepEqual(output, tc.Output) {
+			t.Fatalf("bad: %d\n\n%#v", i, output)
```

Comment:
So I've been thinking about it in each instance, and the main thing for is "does the actual test case provide information vs. what I can provide here" and in this case I figured that the `tc.Input` and `output` give you everything you need to fix the bug. If the test _itself_ is wrong then yes, you'd have to count the indices, but I'm assuming the test case itself is correct and won't need to be changed and that if you're fixing an implementation bug then this would catch it.

So, I'd say in this case, option 2 is best.